web/src/composebox_typeahead.ts

```diff
@@ -642,27 +736,84 @@ export function get_sorted_filtered_items(query, input_element) {
 
     const big_results = get_candidates(query, input_element);
 
-    if (!big_results) {
-        return false;
+    if (big_results === false) {
+        return [];
     }
 
-    const opts = get_stream_topic_data(input_element);
-
-    if (completing === "mention" || completing === "silent_mention") {
-        return filter_and_sort_mentions(big_results.is_silent, token, opts);
+    // These are sorted separately
+    if (big_results.type === "mention") {
+        return big_results.list;
     }
 
-    return filter_and_sort_candidates(completing, big_results, token);
+    assert(big_results.type === completing);
+    return filter_and_sort_candidates(big_results, token);
 }
 
-export function filter_and_sort_candidates(completing, candidates, token) {
-    const matcher = compose_content_matcher(completing, token);
+type PillItem =
+    | UserGroupPillData
+    | UserOrMentionPillData
+    | TopicJumpPillData
+    | TimeJumpPillData
+    | LanguagePillData
+    | TopicPillData
+    | EmojiPillData
+    | SlashCommandPillData
+    | StreamPillData;
+
+type TopicJumpList = {
+    type: "topic_jump";
+    list: TopicJumpPillData[];
+};
+
+type TimeJumpList = {
+    type: "time_jump";
+    list: TimeJumpPillData[];
+};
+
+type LanguageList = {
+    type: "syntax";
+    list: LanguagePillData[];
+};
+
+type TopicList = {
+    type: "topic_list";
+    list: TopicPillData[];
+};
+
+type EmojiList = {
+    type: "emoji";
+    list: EmojiPillData[];
+};
+
+type SlashCommandList = {
+    type: "slash";
+    list: SlashCommandPillData[];
+};
 
-    const small_results = candidates.filter((item) => matcher(item));
+type StreamList = {
+    type: "stream";
+    list: StreamPillData[];
+};
 
-    const sorted_results = sort_results(completing, small_results, token);
+export function filter_and_sort_candidates(
+    candidates:
+        | TopicJumpList
+        | TimeJumpList
+        | LanguageList
+        | TopicList
+        | EmojiList
+        | SlashCommandList
+        | StreamList,
+    token: string,
+): PillItem[] {
+    const filtered_candidates = candidates;
+    // TODO(PR): how to get Typescript to understand the result of filter
+    // is the same type?
```

Comment:
A question for a reviewer -- is there a way to get TypeScript to treat the result of `filter` as the same type as what I passed in, even if the original type is a union of list types?

Reply:
discussed here: https://chat.zulip.org/#narrow/stream/6-frontend/topic/typing.20filter.20on.20union.20of.20list.20types/near/1784425